### PR TITLE
WIP: adds support for pushing to GCR via `mvn deploy`. Has failing in…

### PR DESCRIPTION
**Pull request by mavenraven**
…tegration test.

Hey @dflemstr, I'm trying to figure out how to debug the failing integration test. I've tried variations of `mvn -Dmaven.failsafe.debug failsafe:integration-test` without any success. I also can't figure out how to run these tests via IntelliJ. It'd be awesome if you could help me figure out the failing test, or just fix it if you happen to know what the problem is. Thanks!

### CODE REVIEW

**Comment by mavenraven** [comment]
Here's the build output
[output.txt](https://github.com/spotify/dockerfile-maven/files/1042547/output.txt)

FWIW, @gimaker ran the tests against this WIP on Linux (I'm using OS X) and he is seeing successful builds.

**Comment by dflemstr** [comment]
Does this rely on the Google Cloud SDK being installed?

Also, Travis runs inside of GCE so that might screw things up

File: plugin/src/main/java/com/spotify/plugin/dockerfile/AbstractDockerMojo.java [hunk 1]
```diff
@@ -378,10 +381,21 @@ protected static String formatImageName(@Nonnull String repository, @Nonnull Str
 
   @Nonnull
   protected DockerClient openDockerClient() throws MojoExecutionException {
+    ContainerRegistryAuthSupplier authSupplier = null;
+    try {
+      authSupplier = ContainerRegistryAuthSupplier.forApplicationDefaultCredentials()
+          .build();
+      getLog().info("Using Google application credentials");
+    } catch (IOException ex) {
```
Comment (dflemstr) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/AbstractDockerMojo.java]: This needs to catch `DockerException` as well, which maybe shouldn't be an unchecked exception to begin with? (I realize that would be a breaking change)

**Comment by mattnworb** [comment]
It might make sense to add a system property that can be set when running tests to disable the automatic google credential loading, like -DdisableGoogleCredentials. 

Seems like the automatic loading would be useful for most developers running the plugin (without them having to opt-in to the behavior), but it needs to be turned off in some cases

**Comment by dflemstr** [comment]
@mattnworb I like the current strategy of attempting to load the credentials and falling back to proceeding without them, it's just that the current code has a bug when trying to catch the resulting exception.

**Comment by dflemstr** [comment]
Maybe I'm misunderstanding the desired behavior for this case?

**Comment by mavenraven** [comment]
@dflemstr did you have any thoughts re: debugging the failing integration test?

**Comment by mattnworb** [comment]
https://github.com/spotify/docker-client/pull/773 will fix the case where the loaded credentials aren't scoped to devstorage

**Comment by mattnworb** [comment]
testLowercase fails for me on [this branch](https://gist.github.com/mattnworb/38c07654fbdbbc7362cf966f9c2408fc) and on [the master branch](https://gist.github.com/mattnworb/3d35ff4391cd6a09256dd800b5c54e58) for me on Mac OS X too, although perhaps it has to do with Docker for Mac rather than the test on OS X itself. 

```
$ mvn -V
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T15:39:06-04:00)
Maven home: /usr/local/Cellar/maven/3.5.0/libexec
Java version: 1.8.0_101, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_101.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.4", arch: "x86_64", family: "mac"

$ docker version
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Tue Mar 28 00:40:02 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Fri Mar 24 00:00:50 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Comment by mattnworb** [comment]
This reminds me of other issues I have seen with helios-solo on Docker For Mac - when MainIT calls [`TemporaryJob.deploy()`](https://github.com/spotify/dockerfile-maven/blob/55a04cc56b5c974a506f517adf57bc32a0bd9fa7/plugin/src/it/advanced/frontend/src/test/java/com/spotify/it/frontend/MainIT.java#L88), the method does not return until helios-testing has successfully "probed" the external port of the container to make sure the container is up and running ok. Helios probes a tcp port mapping [by checking to see if it can open a socket to the port](https://github.com/spotify/helios/blob/89a1deba9dd303015e03ac310e488962c9b29e05/helios-testing/src/main/java/com/spotify/helios/testing/DefaultProber.java#L101).

I believe that on DFM, the external port is connect-able immediately after the container is created, even if the container is not yet listening on the internal port on the other side of the mapping.

The fix for this is to add a healthcheck to the TemporaryJob so that can helios can do a higher-level check (can it get a HTTP 2xx response from the port) rather than a tcp-level check.

see https://github.com/spotify/helios/issues/916#issuecomment-221090878

**Comment by mattnworb** [comment]
if you bump docker-client to 8.6.2 it should address the other issue mentioned above

**Comment by mattnworb** [comment]
realized I can commit to the branch directly so now the tests are green ✅  😄

File: plugin/src/main/java/com/spotify/plugin/dockerfile/AbstractDockerMojo.java (hunk 1 above)
Comment (mattnworb) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/AbstractDockerMojo.java]: the two methods in the `try` here can only throw `IOException`:

https://github.com/spotify/docker-client/blob/3a9f04d4d81dd2cfacf8f5bd139f007964d4bac6/src/main/java/com/spotify/docker/client/gcr/ContainerRegistryAuthSupplier.java#L99

https://github.com/spotify/docker-client/blob/3a9f04d4d81dd2cfacf8f5bd139f007964d4bac6/src/main/java/com/spotify/docker/client/gcr/ContainerRegistryAuthSupplier.java#L160

it is the methods called at build-/pull-time that might throw `DockerException`

**Comment by mavenraven** [comment]
@mattnworb fwiw I'm using Docker Machine, not Docker for Mac

**Comment by mattnworb** [comment]
@mavenraven can you rebase / resolve the "branch is out of sync" warning? I confirmed with @dflemstr offline that this looks good now.

**Pull request merged**